marlowe-scripts/src/Language/Marlowe/Scripts/OpenRole.hs

```diff
@@ -0,0 +1,263 @@
+   party as the locked role token.
+
+   In the current version we assume that thread token has the same currency as the role token which can be actually undesired.
+
+   Error codes:
+
+   "1" - Own input (never happen) or Marlowe input not found.
+   "2" - Invalid own value - we expect only the role token(s) and min ADA.
+   "3" - Invalid Marlowe redeemer.
+   "4" - Invalid datum.
+   "5" - Missing thread token.
+-}
+mkOpenRoleValidator
+  :: ValidatorHash
+  -- ^ The hash of the corresponding Marlowe validator.
+  -> V1.TokenName
+  -- ^ Datum should be a thread token name.
+  -> BuiltinData
+  -- ^ We ignore redeemer - no need for decoding
+  -> SubScriptContext
+  -- ^ The script context.
+  -> Bool
+mkOpenRoleValidator
+  marloweValidatorHash
+  threadTokenName
+  _
+  SubScriptContext
+    { scriptContextTxInfo = SubTxInfo{txInfoInputs, txInfoRedeemers}
+    , scriptContextPurpose = Spending txOutRef
+    } = do
+    let marloweValidatorAddress = scriptHashAddress marloweValidatorHash
+        -- Performance:
+        -- In the case of three inputs `find` seems to be faster than custom single pass over the list.
+        -- Inlined pattern matching over `Maybe` in both cases also seems to be faster than separate helper function.
+        ownInput = case find (\TxInInfo{txInInfoOutRef} -> txInInfoOutRef == txOutRef) txInfoInputs of
+          Just input -> input
+          Nothing -> traceError "1" -- Own input not found.
+        marloweInput = case find (\TxInInfo{txInInfoResolved} -> txOutAddress txInInfoResolved == marloweValidatorAddress) txInfoInputs of
+          Just input -> input
+          Nothing -> traceError "1" -- Marlowe input not found.
+        TxInInfo{txInInfoResolved = TxOut{txOutValue = ownValue}} = ownInput
+
+        -- Extract role token information from the own input `Value`.
+        (currencySymbol, roleName) = do
+          let valuesList = AssocMap.toList $ getValue ownValue
+          -- Value should contain only min. ADA and a specific role token(s) (we can have few coins of the same role
+          -- token - they are all released).
+          -- Performance: `find` performs here clearly worse.
+          case valuesList of
+            [(possibleAdaSymbol, _), (currencySymbol, AssocMap.toList -> [(roleName, _)])]
+              | possibleAdaSymbol PlutusTxPrelude.== adaSymbol -> (currencySymbol, roleName)
+            [(currencySymbol, AssocMap.toList -> [(roleName, _)]), _] -> (currencySymbol, roleName)
+            _ -> traceError "2" -- Invalid value - we expect only the role token(s).
+
+        -- In order to release the role token we have to encounter an action which uses/unlocks the role.
+        -- All the other actions will be checked by Marlowe validator itself.
+        marloweRedeemerOk = do
+          let TxInInfo{txInInfoOutRef = marloweTxOutRef} = marloweInput
+              {-# INLINEABLE inputContentUsesRole #-}
```

Comment:
I think these nested functions are automatically inlinable, so the GHC directive is not needed here. The PlutusTx compiler would fail with an error if anything that it needs is not inlinable.